### DISCUSSION TITLE
Is it possible to connect to Dataverse using go-sqlcmd?

### DISCUSSION THREAD

**Discussion by Dzieni** [microsoft/go-sqlcmd#510] [Q&A] [answered]
So Microsoft provides [an SQL compatibility layer to Dataverse](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/dataverse-sql-query).

I connect successfully using SSMS on a Windows VM, but I'm looking for something more lightweight, that I could run natively on macOS. go-sqlcmd seems to be a perfect match.

However, my attempts are not successful:

```
michal@dzieni ~ % sqlcmd -S myorg.api.crm4.dynamics.com     
TLS Handshake failed: unexpected packet 4, expecting prelogin
TLS Handshake failed: unexpected packet 4, expecting prelogin
```

Does it happen because:
- I didn't provide any authentication data yet
- something else PEBKAC-related
- or it's rather a missing feature of sqlcmd?

**Comment by shueybubbles** [comment]
thx for opening an issue!
there could be a subtle discrepancy in the network protocols between real SQL Server and Dataverse that the driver doesn't handle correctly.

try 
`sqlcmd -S myorg.api.crm4.dynamics.com --driver-logging-level 255` and provide the output here.

**Reply by Dzieni** [reply]
There you go :)

```
michal@dzieni ~ % sqlcmd -S myorg.api.crm4.dynamics.com --driver-logging-level 255
DRIVER:Dialing with protocol tcp
DRIVER:Returning connection from protocol tcp
TLS Handshake failed: unexpected packet 4, expecting prelogin
TLS Handshake failed: unexpected packet 4, expecting prelogin
```

**Comment by shueybubbles** [comment]
Supposedly, the `tls` package defaults to TLS version 1.2, but my net sniff shows sqlcmd using TLS 1.0.
I suspect dataverse is blocking TLS 1.0 connections. Need to see if this is a go-mssqldb driver issue.

**Comment by shueybubbles** [comment]
Hmm. I don't think it's a TLS issue. It might be that dataverses uses an older TDS protocol. The Go driver expects something newer than 7.2.
The documentation for TDS points out that pre-7.2, the server would return packet type 4 instead of 18 in response to prelogin:

```
This applies to SSL traffic. The client sends the SSL handshake payloads as data in a PRELOGIN message. For TDS versions earlier than TDS 7.2, the server SHOULD send the SSL handshake payloads as data in a [table response](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/26cd55fe-51c4-4948-bd6b-552f861e9427#gt_71dd1dd2-c167-49a8-a5f5-6b0df5c8b48a) message (0x04). For TDS 7.2, TDS 7.3, and TDS 7.4, the server SHOULD send the SSL handshake payloads as data in a PRELOGIN message. Upon successful completion of the SSL handshake, the client will proceed to send the LOGIN7 stream to the server to initiate authentication.
```

**Comment by shueybubbles** [comment, marked as answer]
Opened https://github.com/microsoft/go-mssqldb/issues/174

**Reply by Dzieni** [reply]
Thanks for following up on this!
